hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -231,6 +231,26 @@ class LowerTableIR(val typesToLower: DArrayLowering.Type) extends AnyVal {
 
         loweredChild.copy(body = StreamFlatMap(loweredChild.body, row.name, StreamMap(fieldRef, elt.name, newRow)))
 
+      case TableRename(child, rowMap, globalMap) =>
+        val loweredChild = lower(child)
+        val structId = genUID()
+
+        val oldRowType = loweredChild.body.typ.asInstanceOf[TStream].elementType.asInstanceOf[TStruct]
+        val newRowType = oldRowType.rename(rowMap)
+
+        val oldGlobalType = loweredChild.broadcastVals.typ.asInstanceOf[TStruct].field(loweredChild.globalsField).typ.asInstanceOf[TStruct]
+        val newGlobalType = oldGlobalType.rename(globalMap)
+
+        val oldBroadcastType = loweredChild.broadcastVals.typ.asInstanceOf[TStruct]
+        val oldBroadcastGlobalIndex = oldBroadcastType.fieldIdx(loweredChild.globalsField)
+        val newBroadcastType = oldBroadcastType.updateKey(loweredChild.globalsField, oldBroadcastGlobalIndex, newGlobalType)
+
+        loweredChild.copy(
+          broadcastVals = CastRename(loweredChild.broadcastVals, newBroadcastType),
```

Comment:
You can't quite do this, since references to the old globals need to use the old names. e.g. 
```
TableMapRows(
    TableRename(
        TableMapRows(
            TableMapGlobals(
                TableRange(10, 3),
                InsertFields(global, IndexedSeq("foo" -> I32(2)))),
            InsertFields(row, FastIndexedSeq("bar" -> GetField(row, "idx") + GetField(global, "foo")))),
        Map(), Map("foo" -> "f")),
    InsertFields(row, FastIndexedSeq("b" -> GetField(row, "bar") + GetField(global, "f"))))
```

I think if we refactored how we handled broadcast values/bindings here so that you could just explicitly reference previously defined values, you could just add a new binding for `CastRename(Ref(oldGlobalsField))`, which would avoid constructing the old globals twice but not broadcasting them twice if you need both the original field and the renamed one, like in the example. I'm not sure if we have another pass that would be able to tell that you were broadcasting the same struct with different names.

Reply:
Makes sense. I think #8627 will refactor the handling of bindings to allow this.